File: JavaScript/120. Triangle.js
/**
 * @param {number[][]} triangle
 * @return {number}
 */
var minimumTotal = function(triangle) {
    var n=triangle.length;
    var dp= Array(n).fill(0);
    for(var i=0; i<n; i++){
        dp[i]=triangle[n-1][i];
    }
    for(var i=n-2; i>=0; i--){
        var arr= Array(n).fill(0);
        for(var j=i; j>=0; j--){
            arr[j]=Math.min(triangle[i][j]+dp[j], triangle[i][j]+dp[j+1]);
            }
        dp=arr;
    }
    return dp[0];
};

File: Java/120. Triangle.java
class Solution {
    public int minimumTotal(List<List<Integer>> triangle) {
        int n=triangle.size();
        int[] dp = new int[n];
        for(int i=0; i<n; i++){
            dp[i]=triangle.get(n-1).get(i);
        }
        for(int i=n-2; i>=0; i--){
            int[] arr = new int[n];
            for(int j=i; j>=0; j--){
                arr[j]=Math.min(triangle.get(i).get(j)+dp[j], triangle.get(i).get(j)+dp[j+1]);
            }
            dp=arr;
        }
        return dp[0];
    }
}
